src/Components/Components/src/PersistentState/PersistentServiceTypeCache.cs

```diff
@@ -0,0 +1,82 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Concurrent;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Components;
+
+// A cache for registered persistent services. This is similar to the RootComponentTypeCache.
+internal sealed class PersistentServiceTypeCache
+{
+    private readonly ConcurrentDictionary<Key, Type?> _typeToKeyLookUp = new();
+
+    public Type? GetPersistentService(string assembly, string type)
+    {
+        var key = new Key(assembly, type);
+        if (_typeToKeyLookUp.TryGetValue(key, out var resolvedType))
+        {
+            return resolvedType;
+        }
+        else
+        {
+            return _typeToKeyLookUp.GetOrAdd(key, ResolveType, AppDomain.CurrentDomain.GetAssemblies());
+        }
+    }
+
+    [UnconditionalSuppressMessage("Trimming", "IL2026", Justification = "Types in this cache are added by calling `AddPersistentService` and are expected to be preserved.")]
+    private static Type? ResolveType(Key key, Assembly[] assemblies)
+    {
+        Assembly? assembly = null;
+        for (var i = 0; i < assemblies.Length; i++)
+        {
+            var current = assemblies[i];
+            if (current.GetName().Name == key.Assembly)
+            {
+                assembly = current;
+                break;
+            }
+        }
+
+        if (assembly == null)
+        {
+            // It might be that the assembly is not loaded yet, this can happen if the root component is defined in a
+            // different assembly than the app and there is no reference from the app assembly to any type in the class
+            // library that has been used yet.
+            // In this case, try and load the assembly and look up the type again.
+            // We only need to do this in the browser because its a different process, in the server the assembly will already
+            // be loaded.
+            if (OperatingSystem.IsBrowser())
+            {
+                try
+                {
+                    assembly = Assembly.Load(key.Assembly);
+                }
+                catch
+                {
+                    // It's fine to ignore the exception, since we'll return null below.
+                }
+            }
+        }
+
+        return assembly?.GetType(key.Type, throwOnError: false, ignoreCase: false);
+    }
+
+    private readonly struct Key : IEquatable<Key>
```

Comment:
can this be a record struct?

Reply:
We avoid record structs where possible because they generate a ton more IL, we use to use them all over the place, but there was a push in the framework to avoid using them where not needed.

Reply:
For reference, this is a "copy" of https://github.com/dotnet/aspnetcore/blob/main/src/Components/Shared/src/RootComponentTypeCache.cs